protocols/server/cloud-events/main.tsp

```diff
@@ -0,0 +1,271 @@
+    @body res: MqttConnectEventFailureResponse;
+};
+
+/**
+ * This event is triggered when the service establishes a Web PubSub CONNECTION(aka. session, not a network connection) with a client.
+ */
+@route("OnMqttClientConnected/Doesn'tMatter/DependOnYourEventHandlerSettings")
+op OnMqttClientConnected(
+    /**
+     * A unique ID generated by the service for each physical connection. Its format may change, and you shouldn't try to parse it.
+     */
+    @header("ce-physicalConnectionId") _: string,
+
+    @bodyRoot
+    req: EmptyConnectedEventRequest & CommonRequestHeaders,
+): {};
+
+/**
+ * This event is triggered when a Web PubSub CONNECTION(aka. session, not a network connection) is ended.
+ */
+@route("OnMqttClientDisconnected/Doesn'tMatter/DependOnYourEventHandlerSettings")
+op OnMqttClientDisconnected(
+    @header contentType: "application/json",
+
+    /**
+     * A unique ID generated by the service for each physical connection. Its format may change, and you shouldn't try to parse it.
+     */
+    @header("ce-physicalConnectionId") _: string,
+
+    @bodyRoot
+    req: MqttDisconnectedEventRequest & CommonRequestHeaders,
+): {};
+
+model EmptyConnectedEventRequest {}
+
+/**
+ * Represents a request for a connection event.
+ */
+model ConnectEventRequest {
+    claims: Record<string[]>;
+    query: Record<string[]>;
+    headers: Record<string[]>;
+    subprotocols: string[];
+    clientCertificates: ClientCertificateInfo[];
+}
+
+model ClientCertificateInfo {
+    thumbprint: string;
+    content: string;
+}
+
+/**
+ * Represents a request for an MQTT connection event.
+ */
+model MqttConnectEventRequest extends ConnectEventRequest {
+    mqtt: MqttProperties;
+}
+
+/**
+ * Represents the properties of an MQTT connection.
+ */
+model MqttProperties {
+    /**
+     * MQTT protocol version. The same as the CONNECT packet's ProtocolVersion.
+     * MQTT 3.1.1 is 4, MQTT 5.0 is 5.
+     */
+    protocolVersion: integer;
+
+    /**
+     * The username field in the MQTT CONNECT packet.
+     */
+    username?: string;
+
+    /**
+     * The base64 encoded password field in the MQTT CONNECT packet.
+     */
+    password?: string;
+
+    /**
+     * The user properties in the MQTT CONNECT packet.
+     */
+    userProperties?: MqttUserProperty[];
+}
+
+/**
+ * Represents a user property in an MQTT connection.
+ */
+model MqttUserProperty {
+    name: string;
+    value: string;
+}
+
+/**
+ * Represents a successful response to an MQTT connection event.
+ */
+model MqttConnectEventSuccessResponse {
+    /**
+     * The property provides a way for the upstream Webhook to authorize the client. There are different roles to grant initial permissions for PubSub WebSocket clients. See details in https://learn.microsoft.com/azure/azure-web-pubsub/concept-client-protocols#permissions
+     */
+    roles?: string[];
+
+    /**
+     * It should be "mqtt" or omitted.
+     */
+    subprotocol?: string;
```

Comment:
maybe directly omit this field if it is mqtt

Reply:
I think it'd be better to have validation other than be flexible.